### PR TITLE
[Rocky8] Add Validator to check CustomAMI is provided when Private OSSes are used

### PR DESCRIPTION
**Pull request by himani2411**
### Description of changes
* Adding Validation for checking if CustomAMI is provided when we have Rocky8 ( Private OSSes) that we dont publicly Publish/release.
* Changed the Integ Test code to get the latest internal Parallelcluster AMI for rocky8.

### Tests
* Local Unit Test
* Custom Resource Integ Test for x86 and ARM instances

### References
* Link to impacted open issues.
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/constants.py [hunk 1]
```diff
@@ -51,6 +51,8 @@
     "rhel8": "rhel8-hvm",
     "rocky8": "rocky8-hvm",
 }
+# We do not publicly publish/release Parallelcluster AMI of below OSSes
+PRIVATE_OSSES = ["rocky8"]
```
Comment (enrico-usai) [inline cli/src/pcluster/constants.py]: let's uses OSES to be consistent with other variables in this document.

File: cli/src/pcluster/config/cluster_config.py [hunk 2]
```diff
@@ -1218,6 +1219,7 @@ def __init__(self, os: str, custom_ami: str = None):
         self.custom_ami = Resource.init_param(custom_ami)
 
     def _register_validators(self, context: ValidatorContext = None):  # noqa: D102 #pylint: disable=unused-argument
+        self._register_validator(PrivateOsCustomAmiValidator, os=self.os, custom_ami=self.custom_ami)
```
Comment (enrico-usai) [inline cli/src/pcluster/config/cluster_config.py]: I'd call it `OsCustomAmiValidator` in this way we can use it for other validations related to os and custom amis in future.

File: cli/src/pcluster/validators/cluster_validators.py [hunk 3]
```diff
@@ -136,6 +137,22 @@ def _validate(self, os, scheduler):
             )
 
 
+class PrivateOsCustomAmiValidator(Validator):
+    """If OS is one of the OSSes that we do not publicly publish/release we need to use a Custom AMI."""
```
Comment (enrico-usai) [inline cli/src/pcluster/validators/cluster_validators.py]: I'd write: "For some OSes we don't publish official AMIs, so CustomAmi parameter is required"

File: cli/src/pcluster/validators/cluster_validators.py [hunk 4]
```diff
@@ -136,6 +137,22 @@ def _validate(self, os, scheduler):
...
+
+    def _validate(self, os: str, custom_ami: str):
+        if not custom_ami and os in PRIVATE_OSSES:
+            self._add_failure(
+                (
+                    # TODO: Add Link
+                    f"Parallelcluster doesnt publish/release official {os} AMI. "
```
Comment (enrico-usai) [inline cli/src/pcluster/validators/cluster_validators.py]: `ParallelCluster` is the correct name of the product.

I'd write: `There is no official AMI for {os}.  Please build your own AMI using `pcluster build-image` command, as explained in the documentation: https://docs.aws.amazon.com/parallelcluster/latest/ug/building-custom-ami-v3.html`

File: tests/integration-tests/conftest.py [hunk 5]
```diff
@@ -684,6 +685,21 @@ def inject_additional_config_settings(cluster_config, request, region, benchmark
 
     if request.config.getoption("custom_ami") and not dict_has_nested_key(config_content, ("Image", "CustomAmi")):
         dict_add_nested_key(config_content, request.config.getoption("custom_ami"), ("Image", "CustomAmi"))
+    # Adding Internal AMI we create for integration testing of those OSses which we do not publicly publish/release
+    if config_content["Image"]["Os"] in PRIVATE_OSSES and not dict_has_nested_key(
+        config_content, ("Image", "CustomAmi")
+    ):
+        dict_add_nested_key(
```
Comment (enrico-usai) [inline tests/integration-tests/conftest.py:692]: Nice approach. Does it permit to run integ tests for released version as well?

Comment (himani2411) [reply tests/integration-tests/conftest.py:692]: > Does it permit to run integ tests for released version as well?

The function [retrieve_lates_ami](https://github.com/aws/aws-parallelcluster/blob/49131fbf848b1beedcbc2110bf233e01c2ebc6f1/tests/integration-tests/tests/common/utils.py#L106-108) will run against a released version of AMI if we dont use any Git refs

Comment (himani2411) [reply tests/integration-tests/conftest.py:692]: I am testing the same by removing the `allow_private_ami` that i had kept as True for testing

Comment (himani2411) [reply tests/integration-tests/conftest.py:692]: Tests successfully

**Review by enrico-usai — APPROVED** [review]

**Pull request merged**